File: client/src/map3d/terrain-generator.js
/**
 * 大陆地形生成器 — 高度图 + 地面网格 + 水面 + 装饰
 *
 * 核心思路：
 * 1. 基于 17 个城市节点位置 + terrain 类型，用 Inverse-Distance Weighting (IDW) 插值生成高度图
 * 2. 顶点着色按最近城市地形类型
 * 3. 水面在 Y < waterLevel 处
 * 4. 地表装饰：岩石/树木/沙纹/水洼/碎石
 */
import * as THREE from 'three';
import { TERRAIN_CONFIG, TERRAIN_GEN, WORLD_SCALE, ELEVATION_SCALE } from './constants.js';
import { toWorldXZ } from './world-space.js';
import { getTerrainTexture } from './terrain-textures.js';

let groundMesh = null;
let waterMesh = null;
let decorationsGroup = null;
let heightData = null; // Float32Array for getHeightAt queries

// 存储节点 3D 位置用于 IDW
let nodePositions3D = []; // [{ x, z, terrain, elevation }]

/**
 * 生成大陆地形
 * @param {object} nodes - stateStore.nodes
 * @returns {{ groundMesh: THREE.Mesh, waterMesh: THREE.Mesh, decorationsGroup: THREE.Group, getHeightAt: Function }}
 */
export function generateTerrain(nodes) {
  const cfg = TERRAIN_GEN;

  // 准备节点 3D 位置
  nodePositions3D = [];
  for (const node of Object.values(nodes)) {
    const { x, z } = toWorldXZ(node.x, node.y);
    const terrainCfg = TERRAIN_CONFIG[node.terrain] || TERRAIN_CONFIG.PLAINS;
    nodePositions3D.push({
      x, z,
      terrain: node.terrain || 'PLAINS',
      elevation: terrainCfg.elevation * ELEVATION_SCALE,
      factionId: node.factionId,
    });
  }

  // 生成高度图
  const { positions, colors, indices, heightMap } = generateHeightMap(cfg);

  // 创建地面网格
  const geometry = new THREE.BufferGeometry();
  geometry.setAttribute('position', new THREE.BufferAttribute(positions, 3));
  geometry.setAttribute('color', new THREE.BufferAttribute(colors, 3));
  geometry.setIndex(indices);
  geometry.computeVertexNormals();

  const material = new THREE.MeshStandardMaterial({
    vertexColors: true,
    roughness: 0.85,
    metalness: 0.05,
    flatShading: false,
  });

  groundMesh = new THREE.Mesh(geometry, material);
  groundMesh.receiveShadow = true;
  groundMesh.name = 'terrain';

  // 水面
  waterMesh = createWaterMesh(cfg);

  // 装饰物
  decorationsGroup = createDecorations(nodes, cfg);

  // 高度查询数据
  heightData = heightMap;

  return {
    groundMesh,
    waterMesh,
    decorationsGroup,
    getHeightAt,
  };
}

/**
 * IDW 高度图生成
 */
function generateHeightMap(cfg) {
  const seg = cfg.gridSize;
  const size = cfg.worldSize;
  const half = size / 2;

  const vertCount = (seg + 1) * (seg + 1);
  const positions = new Float32Array(vertCount * 3);
  const colors = new Float32Array(vertCount * 3);
  const heightMap = new Float32Array(vertCount);

  // 为每个顶点计算高度和颜色
  for (let iz = 0; iz <= seg; iz++) {
    for (let ix = 0; ix <= seg; ix++) {
      const idx = iz * (seg + 1) + ix;
      const wx = -half + (ix / seg) * size;
      const wz = -half + (iz / seg) * size;

      // IDW 插值高度
      let height = 0;
      let weightSum = 0;
      let dominantTerrain = 'PLAINS';
      let maxWeight = 0;

      for (const node of nodePositions3D) {
        const dx = wx - node.x;
        const dz = wz - node.z;
        const dist = Math.sqrt(dx * dx + dz * dz);

        if (dist < 0.01) {
          height = node.elevation;
          weightSum = 1;
          dominantTerrain = node.terrain;
          maxWeight = 1;
          break;
        }

        const w = 1 / Math.pow(dist, cfg.heightBlendPower);
        height += w * node.elevation;
        weightSum += w;

        if (w > maxWeight) {
          maxWeight = w;
          dominantTerrain = node.terrain;
        }
      }

      if (weightSum > 0) height /= weightSum;

      // 添加噪声让地形更自然
      height += simplexNoise(wx * 0.3, wz * 0.3) * 0.3;

      heightMap[idx] = height;

      const vi = idx * 3;
      positions[vi] = wx;
      positions[vi + 1] = height;
      positions[vi + 2] = wz;

      // 顶点着色
      const terrainCfg = TERRAIN_CONFIG[dominantTerrain] || TERRAIN_CONFIG.PLAINS;
      const col = new THREE.Color(terrainCfg.color);
      // 添加轻微随机变化
      const variation = simplexNoise(wx * 0.5, wz * 0.5) * 0.05;
      colors[vi] = Math.max(0, Math.min(1, col.r + variation));
      colors[vi + 1] = Math.max(0, Math.min(1, col.g + variation));
      colors[vi + 2] = Math.max(0, Math.min(1, col.b + variation));
    }
  }

  // 索引（使用普通数组，让 Three.js setIndex 自动包装为 BufferAttribute）
  const indices = [];
  for (let iz = 0; iz < seg; iz++) {
    for (let ix = 0; ix < seg; ix++) {
      const a = iz * (seg + 1) + ix;
      const b = a + 1;
      const c = a + (seg + 1);
      const d = c + 1;
      indices.push(a, c, b);
      indices.push(b, c, d);
    }
  }

  return { positions, colors, indices, heightMap };
}

/**
 * 创建水面
 */
function createWaterMesh(cfg) {
  const size = cfg.worldSize;
  const waterGeo = new THREE.PlaneGeometry(size * 1.5, size * 1.5);
  waterGeo.rotateX(-Math.PI / 2);
  const waterMat = new THREE.MeshStandardMaterial({
    color: cfg.waterColor,
    transparent: true,
    opacity: cfg.waterOpacity,
    roughness: 0.1,
    metalness: 0.3,
    side: THREE.DoubleSide,
  });
  const mesh = new THREE.Mesh(waterGeo, waterMat);
  mesh.position.y = cfg.waterLevel;
  mesh.name = 'water';
  return mesh;
}

/**
 * 创建地表装饰（岩石、树木等）
 */
function createDecorations(nodes, cfg) {
  const group = new THREE.Group();
  group.name = 'decorations';

  const density = cfg.decorationDensity;

  for (const node of nodePositions3D) {
    const count = density[node.terrain] || 0;
    if (count <= 0) continue;

    for (let i = 0; i < count; i++) {
      // 在节点周围随机位置
      const angle = Math.random() * Math.PI * 2;
      const radius = 1 + Math.random() * (CITY_INFLUENCE_RADIUS || 5) * 0.8;
      const wx = node.x + Math.cos(angle) * radius;
      const wz = node.z + Math.sin(angle) * radius;
      const wy = getHeightAt(wx, wz);

      let deco;
      switch (node.terrain) {
        case 'MOUNTAIN':
          deco = createRock(wx, wy, wz);
          break;
        case 'FOREST':
          deco = createTree(wx, wy, wz);
          break;
        case 'DESERT':
          deco = createSandDune(wx, wy, wz);
          break;
        case 'SWAMP':
          deco = createPuddle(wx, wy, wz);
          break;
        case 'RUINS':
          deco = createRubble(wx, wy, wz);
          break;
        default:
          continue;
      }
      if (deco) group.add(deco);
    }
  }

  return group;
}

// ─── 装饰物工厂函数 ───

function createRock(x, y, z) {
  const size = 0.15 + Math.random() * 0.3;
  const geo = new THREE.DodecahedronGeometry(size, 0);
  const mat = new THREE.MeshStandardMaterial({
    color: 0x808080,
    roughness: 0.9,
    metalness: 0.1,
  });
  const mesh = new THREE.Mesh(geo, mat);
  mesh.position.set(x, y + size * 0.4, z);
  mesh.rotation.set(Math.random(), Math.random(), Math.random());
  mesh.castShadow = true;
  return mesh;
}

function createTree(x, y, z) {
  const group = new THREE.Group();
  // 树干
  const trunkH = 0.3 + Math.random() * 0.3;
  const trunk = new THREE.Mesh(
    new THREE.CylinderGeometry(0.03, 0.04, trunkH, 5),
    new THREE.MeshStandardMaterial({ color: 0x5c4033 })
  );
  trunk.position.y = trunkH / 2;
  group.add(trunk);
  // 树冠
  const crownH = 0.3 + Math.random() * 0.2;
  const crown = new THREE.Mesh(
    new THREE.ConeGeometry(0.15 + Math.random() * 0.1, crownH, 6),
    new THREE.MeshStandardMaterial({ color: 0x2d5a1e })
  );
  crown.position.y = trunkH + crownH / 2;
  group.add(crown);
  group.position.set(x, y, z);
  group.castShadow = true;
  return group;
}

function createSandDune(x, y, z) {
  const size = 0.2 + Math.random() * 0.2;
  const geo = new THREE.SphereGeometry(size, 6, 4, 0, Math.PI * 2, 0, Math.PI / 2);
  const mat = new THREE.MeshStandardMaterial({ color: 0xd4a855, roughness: 1.0 });
  const mesh = new THREE.Mesh(geo, mat);
  mesh.position.set(x, y, z);
  mesh.scale.y = 0.4;
  return mesh;
}

function createPuddle(x, y, z) {
  const size = 0.3 + Math.random() * 0.3;
  const geo = new THREE.CircleGeometry(size, 8);
  geo.rotateX(-Math.PI / 2);
  const mat = new THREE.MeshStandardMaterial({
    color: 0x2a4a4a, transparent: true, opacity: 0.5, roughness: 0.1, metalness: 0.2,
  });
  const mesh = new THREE.Mesh(geo, mat);
  mesh.position.set(x, y + 0.02, z);
  return mesh;
}

function createRubble(x, y, z) {
  const size = 0.1 + Math.random() * 0.15;
  const geo = new THREE.CylinderGeometry(size * 0.6, size, size * 2, 5);
  const mat = new THREE.MeshStandardMaterial({ color: 0x6b5b4f, roughness: 0.95 });
  const mesh = new THREE.Mesh(geo, mat);
  mesh.position.set(x, y + size, z);
  mesh.rotation.z = (Math.random() - 0.5) * 0.8;
  return mesh;
}

/**
 * 查询地面高度（双线性插值）
 * @param {number} wx - 世界 X 坐标
 * @param {number} wz - 世界 Z 坐标
 * @returns {number} 地面高度 Y
 */
export function getHeightAt(wx, wz) {
  if (!heightData) return 0;

  const cfg = TERRAIN_GEN;
  const seg = cfg.gridSize;
  const size = cfg.worldSize;
  const half = size / 2;

  // 世界坐标 → 网格坐标
  const gx = ((wx + half) / size) * seg;
  const gz = ((wz + half) / size) * seg;

  // 超出范围返回 0
  if (gx < 0 || gx >= seg || gz < 0 || gz >= seg) return 0;

  const ix = Math.floor(gx);
  const iz = Math.floor(gz);
  const fx = gx - ix;
  const fz = gz - iz;

  const ix1 = Math.min(ix + 1, seg);
  const iz1 = Math.min(iz + 1, seg);

  const w = seg + 1;
  const h00 = heightData[iz * w + ix] || 0;
  const h10 = heightData[iz * w + ix1] || 0;
  const h01 = heightData[iz1 * w + ix] || 0;
  const h11 = heightData[iz1 * w + ix1] || 0;

  // 双线性插值
  const h0 = h00 * (1 - fx) + h10 * fx;
  const h1 = h01 * (1 - fx) + h11 * fx;
  return h0 * (1 - fz) + h1 * fz;
}

/**
 * 简单 2D simplex-like 噪声（用于地形自然变化）
 */
function simplexNoise(x, y) {
  const n = Math.sin(x * 12.9898 + y * 78.233) * 43758.5453;
  return (n - Math.floor(n)) * 2 - 1;
}

// 需要从 constants 导入
const CITY_INFLUENCE_RADIUS = 5.0;

File: U3D/open-world-game-unity/Assets/Scripts/Rendering/TerrainGenerator.cs
using System;
using System.Collections.Generic;
using GameState;
using UnityEngine;
using Random = UnityEngine.Random;

/// <summary>
/// 大陆地形生成器 — IDW 高度图 + 顶点色网格 + 水面
/// 移植自 client/src/map3d/terrain-generator.js (362 行)
/// </summary>
namespace Rendering
{
    public class TerrainGenerator : MonoBehaviour
    {
        [Header("生成参数")]
        [SerializeField] private int _gridSize = 200;
        [SerializeField] private float _worldSize = 60f;

        [Header("材质")]
        [SerializeField] private Material _terrainMaterial;
        [SerializeField] private Material _waterMaterial;

        private Mesh _groundMesh;
        private Mesh _waterMesh;
        private GameObject _groundObject;
        private GameObject _waterObject;
        private float[] _heightData;

        // 节点 3D 位置用于 IDW
        private struct NodeTerrainInfo
        {
            public float x, z;
            public string terrain;
            public float elevation;
            public string factionId;
        }
        private List<NodeTerrainInfo> _nodePositions3D = new();

        /// <summary>生成大陆地形</summary>
        public void Generate(Dictionary<string, NodeComponent> nodes)
        {
            Cleanup();

            if (nodes == null || nodes.Count == 0)
            {
                Debug.LogWarning("[Terrain] 节点数据为空，跳过地形生成");
                return;
            }

            // 准备节点 3D 位置
            PrepareNodePositions(nodes);

            // 生成高度图网格
            var (mesh, heightMap) = GenerateHeightMap();
            _heightData = heightMap;

            // 地面
            _groundMesh = mesh;
            _groundObject = new GameObject("Terrain");
            _groundObject.transform.SetParent(transform, false);
            var mf = _groundObject.AddComponent<MeshFilter>();
            mf.sharedMesh = _groundMesh;
            var mr = _groundObject.AddComponent<MeshRenderer>();
            mr.sharedMaterial = _terrainMaterial ?? CreateDefaultTerrainMaterial();
            _groundObject.isStatic = true;

            // 水面
            _waterMesh = CreateWaterMesh();
            _waterObject = new GameObject("Water");
            _waterObject.transform.SetParent(transform, false);
            var wf = _waterObject.AddComponent<MeshFilter>();
            wf.sharedMesh = _waterMesh;
            var wr = _waterObject.AddComponent<MeshRenderer>();
            wr.sharedMaterial = _waterMaterial ?? CreateDefaultWaterMaterial();
        }

        private void PrepareNodePositions(Dictionary<string, NodeComponent> nodes)
        {
            _nodePositions3D.Clear();

            foreach (var (id, node) in nodes)
            {
                var xz = CoordinateUtils.ToWorldXZ(node.X, node.Y);
                float elevation = GetTerrainElevation(node.Terrain);
                _nodePositions3D.Add(new NodeTerrainInfo
                {
                    x = xz.x,
                    z = xz.y,
                    terrain = node.Terrain ?? "PLAINS",
                    elevation = elevation * Constants.ElevationScale,
                    factionId = node.FactionId,
                });
            }
        }

        private (Mesh mesh, float[] heightMap) GenerateHeightMap()
        {
            int seg = _gridSize;
            float size = _worldSize;
            float half = size / 2f;

            int vertCount = (seg + 1) * (seg + 1);
            var vertices = new Vector3[vertCount];
            var colors = new Color[vertCount];
            var heightMap = new float[vertCount];
            var indices = new List<int>(seg * seg * 6);

            for (int iz = 0; iz <= seg; iz++)
            {
                for (int ix = 0; ix <= seg; ix++)
                {
                    int idx = iz * (seg + 1) + ix;
                    float wx = -half + (float)ix / seg * size;
                    float wz = -half + (float)iz / seg * size;

                    // IDW 插值高度
                    float height = 0f;
                    float weightSum = 0f;
                    string dominantTerrain = "PLAINS";
                    float maxWeight = 0f;

                    foreach (var node in _nodePositions3D)
                    {
                        float dx = wx - node.x;
                        float dz = wz - node.z;
                        float dist = Mathf.Sqrt(dx * dx + dz * dz);

                        if (dist < 0.01f)
                        {
                            height = node.elevation;
                            weightSum = 1f;
                            dominantTerrain = node.terrain;
                            maxWeight = 1f;
                            break;
                        }

                        float w = 1f / Mathf.Pow(dist, Constants.TerrainGen.HeightBlendPower);
                        height += w * node.elevation;
                        weightSum += w;

                        if (w > maxWeight)
                        {
                            maxWeight = w;
                            dominantTerrain = node.terrain;
                        }
                    }

                    if (weightSum > 0f) height /= weightSum;

                    // 添加噪声
                    height += SimplexNoise(wx * 0.3f, wz * 0.3f) * 0.3f;

                    heightMap[idx] = height;

                    // 顶点位置
                    vertices[idx] = new Vector3(wx, height, wz);

                    // 顶点颜色
                    Color terrainColor = GetTerrainColor(dominantTerrain);
                    float variation = SimplexNoise(wx * 0.5f, wz * 0.5f) * 0.05f;
                    colors[idx] = new Color(
                        Mathf.Clamp01(terrainColor.r + variation),
                        Mathf.Clamp01(terrainColor.g + variation),
                        Mathf.Clamp01(terrainColor.b + variation)
                    );
                }
            }

            // 索引
            for (int iz = 0; iz < seg; iz++)
            {
                for (int ix = 0; ix < seg; ix++)
                {
                    int a = iz * (seg + 1) + ix;
                    int b = a + 1;
                    int c = a + (seg + 1);
                    int d = c + 1;
                    indices.Add(a); indices.Add(c); indices.Add(b);
                    indices.Add(b); indices.Add(c); indices.Add(d);
                }
            }

            // 构建 Mesh
            var mesh = new Mesh();
            mesh.indexFormat = vertCount > 65535 ? UnityEngine.Rendering.IndexFormat.UInt32 : UnityEngine.Rendering.IndexFormat.UInt16;
            mesh.vertices = vertices;
            mesh.colors = colors;
            mesh.SetIndices(indices.ToArray(), MeshTopology.Triangles, 0);
            mesh.RecalculateNormals();
            mesh.RecalculateBounds();
            mesh.name = "TerrainHeightmap";

            return (mesh, heightMap);
        }

        private Mesh CreateWaterMesh()
        {
            float size = _worldSize * 1.5f;
            var mesh = new Mesh();
            var verts = new Vector3[]
            {
                new Vector3(-size / 2f, Constants.TerrainGen.WaterLevel, -size / 2f),
                new Vector3( size / 2f, Constants.TerrainGen.WaterLevel, -size / 2f),
                new Vector3(-size / 2f, Constants.TerrainGen.WaterLevel,  size / 2f),
                new Vector3( size / 2f, Constants.TerrainGen.WaterLevel,  size / 2f),
            };
            var tris = new int[] { 0, 2, 1, 1, 2, 3 };
            mesh.vertices = verts;
            mesh.SetIndices(tris, MeshTopology.Triangles, 0);
            mesh.RecalculateNormals();
            mesh.name = "Water";
            return mesh;
        }

        /// <summary>查询地面高度（双线性插值）</summary>
        public float GetHeightAt(float wx, float wz)
        {
            if (_heightData == null || _heightData.Length == 0) return 0f;

            int seg = _gridSize;
            float size = _worldSize;
            float half = size / 2f;

            // 世界坐标 → 网格坐标
            float gx = ((wx + half) / size) * seg;
            float gz = ((wz + half) / size) * seg;

            if (gx < 0 || gx >= seg || gz < 0 || gz >= seg) return 0f;

            int ix = Mathf.FloorToInt(gx);
            int iz = Mathf.FloorToInt(gz);
            float fx = gx - ix;
            float fz = gz - iz;

            int ix1 = Mathf.Min(ix + 1, seg);
            int iz1 = Mathf.Min(iz + 1, seg);

            int w = seg + 1;
            float h00 = _heightData[iz * w + ix];
            float h10 = _heightData[iz * w + ix1];
            float h01 = _heightData[iz1 * w + ix];
            float h11 = _heightData[iz1 * w + ix1];

            float h0 = h00 * (1f - fx) + h10 * fx;
            float h1 = h01 * (1f - fx) + h11 * fx;
            return h0 * (1f - fz) + h1 * fz;
        }

        /// <summary>简化 2D 伪随机噪声</summary>
        private static float SimplexNoise(float x, float y)
        {
            float n = Mathf.Sin(x * 12.9898f + y * 78.233f) * 43758.5453f;
            return (n - Mathf.Floor(n)) * 2f - 1f;
        }

        /// <summary>根据地形类型获取海拔</summary>
        private static float GetTerrainElevation(string terrainType)
        {
            foreach (var t in Constants.TerrainConfig)
            {
                if (t.Name == terrainType) return t.Elevation;
            }
            return 0f;
        }

        /// <summary>根据地形类型获取颜色</summary>
        private static Color GetTerrainColor(string terrainType)
        {
            foreach (var t in Constants.TerrainConfig)
            {
                if (t.Name == terrainType)
                    return ColorFromInt(t.Color);
            }
            return ColorFromInt(Constants.TerrainConfig[0].Color);
        }

        /// <summary>int 颜色 → Color</summary>
        private static Color ColorFromInt(int hex)
        {
            float r = ((hex >> 16) & 0xFF) / 255f;
            float g = ((hex >> 8) & 0xFF) / 255f;
            float b = (hex & 0xFF) / 255f;
            return new Color(r, g, b);
        }

        private static Material CreateDefaultTerrainMaterial()
        {
            var mat = new Material(Shader.Find("Universal Render Pipeline/Lit"));
            if (mat != null)
            {
                mat.EnableKeyword("_VERTEX_COLORS");
                mat.SetFloat("_Glossiness", 0f);
            }
            return mat;
        }

        private static Material CreateDefaultWaterMaterial()
        {
            var mat = new Material(Shader.Find("Universal Render Pipeline/Lit"));
            if (mat != null)
            {
                mat.color = new Color(0x1E / 255f, 0x90 / 255f, 0xFF / 255f, 0.6f);
                mat.SetFloat("_Surface", 1f); // Transparent
                mat.SetFloat("_Blend", 0f);
                mat.SetFloat("_Glossiness", 0.3f);
                mat.SetFloat("_Metallic", 0.3f);
            }
            return mat;
        }

        /// <summary>清理所有地形对象</summary>
        public void Cleanup()
        {
            if (_groundObject != null) DestroyImmediate(_groundObject);
            if (_waterObject != null) DestroyImmediate(_waterObject);
            if (_groundMesh != null) DestroyImmediate(_groundMesh);
            if (_waterMesh != null) DestroyImmediate(_waterMesh);
            _heightData = null;
            _nodePositions3D.Clear();
        }

        private void OnDestroy() => Cleanup();
    }
}
